_posts/2022-04-01-First-steps-with-swift-concurrency.md

```diff
@@ -0,0 +1,341 @@
+
+Now we create the `AsyncStream` using its intializer, which is returning the `continuation` as a closure parameter, which we will use later on (more about [AsyncStream](https://developer.apple.com/documentation/swift/asyncstream)). In short continuation is needed to inform the stream about the received elements or when the stream is finishing its work.
+
+Please note that `func getThumbnails()` is not an `async` function, because we would like to only bind the data and receive them later on. That's why all async code was wrapped as a `Task`. The `nonisolated` prefix gave a possibility for the function to synchronously create the `AsyncStream` in the code. Since we do not modify anything in terms of the actors properties in the function, it is safe to expose the method as `nonisolated`.
+
+And then the usage of the Task Group looks like below:
+
+```swift
+private func fetchThumbnails(
+        from apody: [ApodModel],
+        continuation: ThumbnailsStream.Continuation
+    ) async throws {
+        try await withThrowingTaskGroup(of: (String, UIImage).self) { group in
+            for apod in apody.filter({ $0.media_type == .image }) {
+                group.addTask(priority: .background) {
+                    (apod.url, try await self.fetchImage(url: apod.url))
+                }
+            }
+            for try await data in group {
+                continuation.yield(data)
+            }
+            continuation.finish()
+        }
+}
+```
+
+Once added a child task to a group (by calling `group.addTask()`), child tasks are being executed immediately and in any order.
+
+The `continuation` property yields the result every time the group task finishes its fetch. Once all of the data are fetched, we call `conitnuation.finish()` to inform that the stream has finished publishing data.
+
+The binding in the parent would look like this:
+
+```swift
+@MainActor class ApodViewModel: ObservableObject {
+ 
+	@Published var thumbnails: [String: UIImage] = [:]
+
+	private let dataSource: ThumbnailDataSource
+	...
+	public func fetchThumbnails(for models: [ApodModel]) async {
+        for await (url, image) in dataSource.getThumbnails(models: models) {
+            thumbnails[url] = image
+        }
+    }
+	...
+}
+```
+
+Great, now every time we call `fetchThumbnails(for models: [ApodModel])`, all of the thumbnails for the passed ApodModels will be fetched in the background, by delivering fetched images one after another and populate the thumbnails dictionary.
+
+As you may notice our ViewModel (`ApodViewModel`) is responsible for fetching thumbnails and updating the UI when new thumbnails appears (via `@Published var thumbnails` property and `ObservableObject` conformance). It is marked as a `@MainActor`, which means that every update of the properties or actions within the functions will be called on the main thread. So in our case when the thumbnails are fetched - the UI will be updated from the main thread.
+
+# SecionedFetchRequest
```

Comment:
```suggestion
# SectionedFetchRequest
```

Reply:
We could do with the small intro about this part of the article. I don't understand why it is here and how it is connected to the previous parts.

Eg. "Now let's focus on an improvement" or "let's do ..." + explanation

Reply:
@barcis89 ;)